# FEATURE - Periodic task scheduler

blackironj opened the pull request:
- scheduler manages periodic task
  - Register , Start, Stop, Cancel(remove)

### How to use 
  - Make a new scheduler ( **NewScheduler** )
  - register tasks ( **RegisterTask** )
  - Run a scheduler ( **Run** )
    - if you want to run specific tasks, Use function like this ( Run("taskName1", "taskName2"))

  detail : test code

- Add Test code

blackironj committed:
# FEATURE - Periodic task scheduler

- scheduler manages periodic task
  - Register a task, Start, Stop, Cancel(remove)

- Add Test code

blackironj commented:
@Hyejin @SeongKwangMoon 
주기적으로 하는 작업에 대해서 좀더 편하게 사용할 수 있도록 `Scheduler` 라는 util을 구현하였습니다.

추가적으로 있었으면 하는 기능이 있거나 고쳐야할 부분 확인 부탁드려요
이게 multi thread 환경에서도 안전할 수 있도록 mutex를 걸긴했는데 잘 했는지 모르겠네요

dreamx-one commented on scheduler/scheduler.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+//RegisterTask regiseter task
+func (s *Scheduler) RegisterTask(interval time.Duration, immediately bool, taskNameKey string, taskFunc interface{}, params ...interface{}) error {
```
socket-io golang 라이브러리를 보다보니까 알게 된건데, 이 부분을 마치 자바스크립트의 callback처럼 구현할 수 있는 방법이 있더라고요. 나중에 시간나면 살펴보세요.
https://github.com/graarh/golang-socketio/blob/2c44953b9b5fcd4da83ec8816f002ce68a12ef2d/caller.go

blackironj replied:
@dreamx-one 
살펴보니까 비슷하게 사용하고있는거 같아요. 

이부분은 말 그대로 Task를 등록만 하는 부분입니다.

실제 일괄적으로 도는 부분은 `Run` 에서 해당 함수를 호출할수 있게해놓았어요.

blackironj commented on scheduler/scheduler.go:
```diff
@@ -0,0 +1,168 @@
+			}
+		}
+	}()
+}
```
@dreamx-one 
이 부분에서 실제 함수를 호출하게 되죠

dreamx-one replied:
함수를 그냥 f interface{} 로 받은 다음에 여기에 refect을 써서 func하고 args를 구별해서 등록해놓고 나중에 호출하면 됩니다.

dreamx-one replied:
refect 사용법은 https://cjwoov.tistory.com/16 요기 잘 설명되어 있네요.

dreamx-one reviewed (approved):
👍

blackironj committed:
# EDIT - pkg name ( scheduler -> periodic )

- pkg name `schduler` is little bit ambiguous so modify to `periodic`

- Add test : anonymous function case
